PennMobile/Dining/SwiftUI/DiningAnalyticsViewModel.swift

```diff
@@ -67,36 +68,46 @@ class DiningAnalyticsViewModel: ObservableObject {
+            let dollarPredictions = self.getPredictions(firstBalance: maxDollarBalance, lastBalance: lastDollarBalance)
+            self.dollarSlope = dollarPredictions.slope
+            self.dollarPredictedZeroDate = dollarPredictions.predictedZeroDate
+            self.predictedDollarSemesterEndBalance = dollarPredictions.predictedEndBalance
+            self.dollarAxisLabel = self.getAxisLabelsYX(from: self.dollarHistory)
+            let swipePredictions = self.getPredictions(firstBalance: maxSwipeBalance, lastBalance: lastSwipeBalance)
+            self.swipeSlope = swipePredictions.slope
+            self.swipesPredictedZeroDate = swipePredictions.predictedZeroDate
+            self.predictedSwipesSemesterEndBalance = swipePredictions.predictedEndBalance
             self.swipeAxisLabel = self.getAxisLabelsYX(from: self.swipeHistory)
-            self.predictedSwipesZeroPoint = PredictionsGraphView.getPredictionZeroPoint(startOfSemester: Date.startOfSemester, endOfSemester: Date.endOfSemester, predictedZeroDate: self.swipesPredictedZeroDate)
         }
     }
-    func predictZeroDate(firstBalance: DiningAnalyticsBalance, lastBalance: DiningAnalyticsBalance) -> Date {
-        let yIntercept = firstBalance.balance
-        let slope = (lastBalance.balance - firstBalance.balance) / Double(Calendar.current.dateComponents([.day], from: firstBalance.date, to: lastBalance.date).day!)
-        let offset = -yIntercept / slope
-        var dateComponent = DateComponents()
-        dateComponent.day = Int(offset)
-        return Calendar.current.date(byAdding: dateComponent, to: firstBalance.date)!
-    }
-    func predictSemesterEndBalance(firstBalance: DiningAnalyticsBalance, lastBalance: DiningAnalyticsBalance) -> Double {
-        // y = slope * x + yIntercept
-        let yIntercept = firstBalance.balance
-        let slope = (lastBalance.balance - firstBalance.balance) / Double(Calendar.current.dateComponents([.day], from: firstBalance.date, to: lastBalance.date).day!)
-        let diffInDays = Calendar.current.dateComponents([.day], from: firstBalance.date, to: Date.endOfSemester).day!
-        return ((slope * Double(diffInDays)) + yIntercept)
+    func getPredictions(firstBalance: DiningAnalyticsBalance, lastBalance: DiningAnalyticsBalance) -> (slope: Double, predictedZeroDate: Date, predictedEndBalance: Double) {
+        if firstBalance.date == lastBalance.date {
+            let zeroDate = Calendar.current.date(byAdding: .day, value: 1, to: Date.endOfSemester)!
+            return (Double.infinity, zeroDate, lastBalance.balance)
+        } else if firstBalance.balance == lastBalance.balance {
+            let zeroDate = Calendar.current.date(byAdding: .day, value: 1, to: Date.endOfSemester)!
+            return (Double(0.0), zeroDate, lastBalance.balance)
+        } else {
+            let yIntercept = firstBalance.balance
```

Comment:
I like that you combined this all into one function to avoid repeated code, but it seems like the repeated code from the old `predictZeroDate` and `predictSemesterEndBalance` functions were just calculating the slope. I think this might be a bit more readable if we kept `predictZeroDate`, `predictSemesterEndBalance`, and write a `getSlope` function. That way, `predictZeroDate` and `predictSemesterEndBalance` can either have slope passed in as a parameter or just call `getSlope` within their own functions. Then,`getPredictions` could look like this:

```
func getPredictions(firstBalance: DiningAnalyticsBalance, lastBalance: DiningAnalyticsBalance) -> (slope: Double, predictedZeroDate: Date, predictedEndBalance: Double) {
  if ____ {
    ...
  }
  var slope = getSlope(firstBalance, lastBalance)
  let zeroDate = predictZeroDate(firstBalance, lastBalance, slope)
  let endBalance = predictSemesterEndBalance(firstBalance, lastBalance, slope)
  return (slope, zeroDate, endBalance) 
}
```

I think it's just a little bit nicer to have the calculations abstracted out.

Reply:
I can definitely separate out the slope calculation and the others so it's more readable. However, I would like to note that currently, the slope used to calculate the zeroDate and the endBalance is different from the slope that's returned from getPredictions. The prior one is needed to calculate the zeroDate, but the returned one is the one that's used for the graph (it's scaled differently). I'm going to look into the exact differences between them and see if I can simplify it before fixing it.

Reply:
I don't really see any way to get the correct slope for the graph from the original slope value (for calculating zeroDate) without already doing what is done, so I left it at that.